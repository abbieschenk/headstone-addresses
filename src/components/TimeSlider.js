import React, {
    useState,
    useRef,
    useEffect,
    useCallback,
} from "react";
import PropTypes from "prop-types"

import "../styles/TimeSlider.scss";

/** Adapted from https://dev.to/sandra_lewis/building-a-multi-range-slider-in-react-from-scratch-4dl1 **/

const TimeSlider = ({ 
    min, 
    max, 
    onMinChange, 
    onMaxChange, 
    className }) => {

    const [minVal, setMinVal] = useState(min);
    const [maxVal, setMaxVal] = useState(max);
    
    const minValRef = useRef(min);
    const maxValRef = useRef(max);

    const range = useRef(null);

    // Convert to percentage
    const getPercent = useCallback(value => Math.round(((value - min) / (max - min)) * 100), [min, max]);

    // Set width of the range to change from the left side
    useEffect(() => {
        const minPercent = getPercent(minVal);
        const maxPercent = getPercent(maxValRef.current);

        if (range.current) {
            range.current.style.left = `${minPercent}%`;
            range.current.style.width = `${maxPercent - minPercent}%`;
        }
    }, [minVal, getPercent]);

    // Set width of the range to change from the right side
    useEffect(() => {
        const minPercent = getPercent(minValRef.current);
        const maxPercent = getPercent(maxVal);

        if (range.current) {
        range.current.style.width = `${maxPercent - minPercent}%`;
        }
    }, [maxVal, getPercent]);

    return(    
        <div className={"timeslider" + (className ? " " + className : "")}>
            <input
                type="range"
                min={min}
                max={max}
                value={minVal}
                onChange={event => 
                    {
                        const value = Math.min(Number(event.target.value), maxVal - 1);
                        setMinVal(value);
                        minValRef.current = value;

                        if(onMinChange) {
                            onMinChange(value);
                        }
                    }
                }
                className="thumb thumb--left"
                style={{ zIndex: minVal > max - 100 && "5" }}
            />
            <input
                type="range"
                min={min}
                max={max}
                value={maxVal}
                onChange={event => 
                    {
                        const value = Math.max(Number(event.target.value), minVal + 1);
                        setMaxVal(value);
                        maxValRef.current = value;
                        
                        if(onMaxChange) {
                            onMaxChange(value);
                        }
                    }
                }
            className="thumb thumb--right"
            />
            <div className="slider">
                <div className="slider__track" />
                <div 
                    ref={range} 
                    className="slider__range" 
                />
                <div className="slider__left-value">Year of Birth {minVal}</div>
                <div className="slider__right-value">Year of Death {maxVal}</div>
            </div>
        </div>
    );
};

// Set each prop as type number
TimeSlider.propTypes = {
    min: PropTypes.number.isRequired,
    max: PropTypes.number.isRequired
};

export default TimeSlider;