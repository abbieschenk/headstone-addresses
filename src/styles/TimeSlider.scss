/** Time Slider **/
/** Adapted from https://dev.to/sandra_lewis/building-a-multi-range-slider-in-react-from-scratch-4dl1 **/

$slider-height: 50px;
$slider-width: 400px;

.slider {
    position: relative;
    width: $slider-width;
}
  
.slider__track,
.slider__range {
    border-radius: 3px;
    height: $slider-height;
    position: absolute;
}
  
.slider__track {
    background-color: #ced4da;
    width: 100%;
    z-index: 1;
}
  
.slider__range {
    background-color: #9fe5a8;
    z-index: 2;
}
  
  /* Removing the default appearance */
.thumb,
.thumb::-webkit-slider-thumb {
    -webkit-appearance: none;
}
  
.thumb {
    pointer-events: none;
    position: absolute;
    height: 0;
    width: $slider-width;
    outline: none;
}
  
.thumb--left {
    z-index: 3;
    margin-left: -1px;
}
  
.thumb--right {
    z-index: 4;
}

@mixin thumb-slider {
    background-color: #f1f5f7;
    border: none;
    border-radius: 0;
    box-shadow: 0 0 1px 1px #ced4da;
    cursor: pointer;
    height: $slider-height + 4px;
    width: 12px;
    margin-top: $slider-height - 3px;
    pointer-events: all;
    position: relative;
}
  
/* For Chrome browsers */
.thumb::-webkit-slider-thumb {
    @include thumb-slider();
}

/* For Firefox browsers */
.thumb::-moz-range-thumb {
    @include thumb-slider();
}

.slider__left-value,
.slider__right-value {
    font-size: 12px;
    margin-top: $slider-height + 5px;
    position: absolute;
}

.slider__left-value {
    left: 6px;
}

.slider__right-value {
    right: -4px;
}