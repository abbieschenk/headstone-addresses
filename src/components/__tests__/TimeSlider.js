import React from "react"
import renderer from "react-test-renderer"

import TimeSlider from "../TimeSlider"

describe("TimeSlider", () => {
    
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <TimeSlider min={1900}
            max={2000}
            onMinChange={(value) => {/* do nothing */}} 
            onMaxChange={(value) => {/* do nothing */}}
            className="active"
            >
                <div>Test</div>
            </TimeSlider>
        ).toJSON();

    expect(tree).toMatchSnapshot();
  });

});