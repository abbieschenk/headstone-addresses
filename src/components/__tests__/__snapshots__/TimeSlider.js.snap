// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeSlider renders correctly 1`] = `
<div
  className="timeslider active"
>
  <input
    className="thumb thumb--left"
    max={2000}
    min={1900}
    onChange={[Function]}
    style={
      Object {
        "zIndex": false,
      }
    }
    type="range"
    value={1900}
  />
  <input
    className="thumb thumb--right"
    max={2000}
    min={1900}
    onChange={[Function]}
    type="range"
    value={2000}
  />
  <div
    className="slider"
  >
    <div
      className="slider__track"
    />
    <div
      className="slider__range"
    />
    <div
      className="slider__left-value"
    >
      Year of Birth 
      1900
    </div>
    <div
      className="slider__right-value"
    >
      Year of Death 
      2000
    </div>
  </div>
</div>
`;
